From 7743f705cc8e5562d727e588ff6ac2d6f14c90c2 Mon Sep 17 00:00:00 2001
From: Evgeny Vereshchagin <evvers@ya.ru>
Date: Wed, 28 Feb 2024 04:25:18 +0000
Subject: [PATCH] avahi-utils: fix use-after-free in avahi-browse

Service resolvers kept outside of the client have to be freed and
removed from the list of resolvers kept in the client before the client
itself is freed to prevent them from being freed twice.

Closes https://github.com/avahi/avahi/issues/534
---
 avahi-utils/avahi-browse.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/avahi-utils/avahi-browse.c b/avahi-utils/avahi-browse.c
index 4101895..4028ca0 100644
--- a/avahi-utils/avahi-browse.c
+++ b/avahi-utils/avahi-browse.c
@@ -584,15 +584,15 @@ static void client_callback(AvahiClient *c, AvahiClientState state, AVAHI_GCC_UN
 
                 fprintf(stderr, _("Disconnected, reconnecting ...\n"));
 
+                while (services)
+                    remove_service(config, services);
+
                 avahi_client_free(client);
                 client = NULL;
 
                 avahi_string_list_free(browsed_types);
                 browsed_types = NULL;
 
-                while (services)
-                    remove_service(config, services);
-
                 browsing = 0;
 
                 if (!(client = avahi_client_new(avahi_simple_poll_get(simple_poll), AVAHI_CLIENT_NO_FAIL, client_callback, config, &error))) {
